fgpyo/util/types.py

```diff
@@ -73,9 +84,52 @@ def is_constructible_from_str(type_: type) -> bool:
     return False
 
 
-def _is_optional(type_: type) -> bool:
-    """Returns true if type_ is optional"""
-    return typing.get_origin(type_) is Union and type(None) in typing.get_args(type_)
+# NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
+TypeAnnotation: TypeAlias = Union[type, typing._GenericAlias, UnionType]  # type: ignore[name-defined]
+"""
+A function parameter's type annotation may be any of the following:
+    1) `type`, when declaring any of the built-in Python types
+    2) `typing._GenericAlias`, when declaring generic collection types or union types using pre-PEP
+        585 and pre-PEP 604 syntax (e.g. `List[int]`, `Optional[int]`, or `Union[int, None]`)
+    3) `types.UnionType`, when declaring union types using PEP604 syntax (e.g. `int | None`)
+    4) `types.GenericAlias`, when declaring generic collection types using PEP 585 syntax (e.g.
+       `list[int]`)
+`types.GenericAlias` is a subclass of `type`, but `typing._GenericAlias` and `types.UnionType` are
+not and must be considered explicitly.
+"""
+
+# TODO When dropping support for Python 3.9, deprecate this in favor of performing instance checks
+# directly on the `TypeAnnotation` union type.
+# NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
+TYPE_ANNOTATION_TYPES = (type, typing._GenericAlias, UnionType)  # type: ignore[attr-defined]
+
+
+def _is_optional(dtype: TypeAnnotation) -> bool:
+    """Check if a type is `Optional`.
+    An optional type may be declared using three syntaxes: `Optional[T]`, `Union[T, None]`, or `T |
+    None`. All of these syntaxes is supported by this function.
+    Args:
+        dtype: A type.
+    Returns:
+        True if the type is a union type with exactly two elements, one of which is `None`.
+        False otherwise.
+    Raises:
+        TypeError: If the input is not a valid `TypeAnnotation` type (see above).
+    """
```

Comment:
```suggestion
    """
    Check if a type is `Optional`.
    
    An optional type may be declared using three syntaxes: `Optional[T]`, `Union[T, None]`, or `T |
    None`. All of these syntaxes are supported by this function.
    
    Args:
        dtype: A type.
        
    Returns:
        True if the type is a union type with exactly two elements, one of which is `None`.
        False otherwise.
        
    Raises:
        TypeError: If the input is not a valid `TypeAnnotation` type. 
        Type annotations may be any of `type`, `types.UnionType`, `types.GenericAlias`, or `typing._GenericAlias`.
    """
```